File: scripts/generate_auth_secrets.py
#!/usr/bin/env python3
"""Generate the three Worker secrets that back cloud sign-in.

Same output as scripts/generate-auth-secrets.mjs -- this Python version
exists because macOS ships python3 already, so it needs nothing
installed. Use whichever you prefer; they produce interchangeable
values.

    python3 scripts/generate_auth_secrets.py

Safety properties, on purpose:
  * getpass() reads the password with echo OFF, so it never appears on
    screen, in shell history, or in a process listing -- which is why
    the password is not accepted as a command-line argument.
  * Nothing is written to disk. The password lives in memory only for as
    long as it takes to derive the hash.
  * Only the SALT and HASH are printed. Neither reveals the password:
    recovering it would mean brute-forcing PBKDF2-SHA256 at 210,000
    iterations per guess.
  * Asks twice and aborts on mismatch. A typo here would produce a hash
    you could never log in against.

The derivation matches pbkdf2Hex() in worker.js exactly: PBKDF2-HMAC-
SHA256, 210,000 iterations, 32-byte output, and the salt hashed as the
UTF-8 bytes of the salt STRING (not decoded hex). Change one and you
must change both, or existing logins stop working.
"""

import hashlib
import secrets
import sys
from getpass import getpass

# Keep in sync with PBKDF2_ITERATIONS in worker.js.
PBKDF2_ITERATIONS = 210_000
MIN_PASSWORD_LENGTH = 12


def derive_password_hash(password: str, salt: str) -> str:
    return hashlib.pbkdf2_hmac(
        "sha256", password.encode("utf-8"), salt.encode("utf-8"), PBKDF2_ITERATIONS, 32
    ).hex()


def main() -> int:
    print("\nCloud sign-in secrets for WJ Safety")
    print("-----------------------------------")
    print("Your password is not shown as you type, and is never saved anywhere.\n")

    if not sys.stdin.isatty():
        print("stdin is not a terminal - run this directly so the password can be "
              "typed without being echoed or stored.", file=sys.stderr)
        return 1

    password = getpass("Choose a password:  ")
    if len(password) < MIN_PASSWORD_LENGTH:
        print(f"\nToo short - use at least {MIN_PASSWORD_LENGTH} characters. This single "
              "password is the only thing protecting the cloud API.", file=sys.stderr)
        return 1

    if password != getpass("Type it again:      "):
        print("\nThe two entries did not match. Nothing was generated - run it again.",
              file=sys.stderr)
        return 1

    salt = secrets.token_hex(16)
    password_hash = derive_password_hash(password, salt)
    session_secret = secrets.token_hex(32)

    # Prove the emitted pair really validates the password, so a broken
    # hash can never reach production and lock you out.
    if derive_password_hash(password, salt) != password_hash:
        print("\nSelf-check failed - refusing to output a hash that may not work.",
              file=sys.stderr)
        return 1
    print("Self-check passed: this salt + hash validates the password you typed.\n")

    print("Set these three values as Worker secrets:\n")
    print("  AUTH_PASSWORD_SALT")
    print(f"      {salt}\n")
    print("  AUTH_PASSWORD_HASH")
    print(f"      {password_hash}\n")
    print("  SESSION_SECRET")
    print(f"      {session_secret}\n")
    print("Notes:")
    print("  * Set them on EVERY environment, including preview - an environment")
    print("    without them keeps the API locked (the app still works offline).")
    print("  * Rotating SESSION_SECRET immediately signs out every device.")
    print("  * Keep the password in a password manager. It cannot be recovered from")
    print("    the hash - if you lose it, re-run this and set a new one.")
    print("  * Clear your terminal afterwards so the values aren't left on screen.\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/generate-auth-secrets.mjs
#!/usr/bin/env node
// ---------------------------------------------------------------------
// generate-auth-secrets.mjs
//
// Generates the three Worker secrets that back cloud sign-in:
//   AUTH_PASSWORD_SALT, AUTH_PASSWORD_HASH, SESSION_SECRET
//
// Run it locally:
//   node scripts/generate-auth-secrets.mjs
//
// Safety properties, on purpose:
//   * The password is read from the terminal with echo OFF, so it never
//     appears on screen, in your shell history, or in a process listing
//     (which is exactly why it is NOT a command-line argument).
//   * Nothing is written to disk. The password exists only in memory for
//     the moment it takes to derive the hash.
//   * Only the SALT and the HASH are printed. Neither reveals the
//     password: recovering it would require brute-forcing PBKDF2-SHA256
//     at 210,000 iterations per guess.
//   * It asks twice and refuses to continue on a mismatch. A typo here
//     would produce a hash you could never log in against.
//
// The derivation below is deliberately identical to pbkdf2Hex() in
// worker.js -- same WebCrypto call, same iteration count, same treatment
// of the salt (the UTF-8 bytes of the salt STRING, not decoded hex). If
// you ever change one, change both, or existing logins stop working.
// ---------------------------------------------------------------------

import { webcrypto, randomBytes } from "node:crypto";

// Keep in sync with PBKDF2_ITERATIONS in worker.js.
const PBKDF2_ITERATIONS = 210000;
const MIN_PASSWORD_LENGTH = 12;

export async function derivePasswordHash(password, saltString) {
  const enc = new TextEncoder();
  const keyMaterial = await webcrypto.subtle.importKey("raw", enc.encode(password), "PBKDF2", false, ["deriveBits"]);
  const bits = await webcrypto.subtle.deriveBits(
    { name: "PBKDF2", salt: enc.encode(saltString), iterations: PBKDF2_ITERATIONS, hash: "SHA-256" },
    keyMaterial,
    256
  );
  return [...new Uint8Array(bits)].map((b) => b.toString(16).padStart(2, "0")).join("");
}

// Reads a line from the TTY without echoing it. Refuses to run when
// stdin is not a terminal (piped/redirected), because in that case the
// password would likely be sitting in a file or a shell history entry --
// better to fail loudly than to quietly accept an exposed secret.
function readSecretLine(prompt) {
  return new Promise((resolve, reject) => {
    if (!process.stdin.isTTY) {
      reject(new Error("stdin is not a terminal — run this directly so the password can be typed without being echoed or stored."));
      return;
    }
    process.stdout.write(prompt);
    process.stdin.setRawMode(true);
    process.stdin.resume();
    process.stdin.setEncoding("utf8");

    let value = "";
    const onData = (chunk) => {
      for (const ch of chunk) {
        if (ch === "\r" || ch === "\n") {
          process.stdin.setRawMode(false);
          process.stdin.pause();
          process.stdin.removeListener("data", onData);
          process.stdout.write("\n");
          resolve(value);
          return;
        }
        if (ch === "\u0003") { // Ctrl+C
          process.stdin.setRawMode(false);
          process.stdout.write("\n");
          process.exit(130);
        }
        if (ch === "\u007f" || ch === "\b") { // backspace
          value = value.slice(0, -1);
          continue;
        }
        if (ch >= " ") value += ch;
      }
    };
    process.stdin.on("data", onData);
  });
}

async function main() {
  console.log("\nCloud sign-in secrets for WJ Safety");
  console.log("───────────────────────────────────");
  console.log("Your password is not shown as you type, and is never saved anywhere.\n");

  const password = await readSecretLine("Choose a password:  ");
  if (password.length < MIN_PASSWORD_LENGTH) {
    console.error(`\nToo short — use at least ${MIN_PASSWORD_LENGTH} characters. This single password is the only thing protecting the cloud API.`);
    process.exit(1);
  }

  const confirm = await readSecretLine("Type it again:      ");
  if (password !== confirm) {
    console.error("\nThe two entries did not match. Nothing was generated — run it again.");
    process.exit(1);
  }

  const salt = randomBytes(16).toString("hex");
  const hash = await derivePasswordHash(password, salt);
  const sessionSecret = randomBytes(32).toString("hex");

  // Prove the stored pair actually validates the password, so a broken
  // hash can never reach production and lock you out.
  if ((await derivePasswordHash(password, salt)) !== hash) {
    console.error("\nSelf-check failed — refusing to output a hash that may not work.");
    process.exit(1);
  }
  console.log("Self-check passed: this salt + hash validates the password you typed.\n");

  console.log("Run these three commands. Paste each value when prompted:\n");
  console.log("  npx wrangler secret put AUTH_PASSWORD_SALT");
  console.log(`      ${salt}\n`);
  console.log("  npx wrangler secret put AUTH_PASSWORD_HASH");
  console.log(`      ${hash}\n`);
  console.log("  npx wrangler secret put SESSION_SECRET");
  console.log(`      ${sessionSecret}\n`);
  console.log("Notes:");
  console.log("  • Set these on EVERY environment, including preview — an environment");
  console.log("    without them keeps the API locked (the app still works offline).");
  console.log("  • Rotating SESSION_SECRET immediately signs out every device.");
  console.log("  • Keep the password in a password manager. It cannot be recovered");
  console.log("    from the hash — if you lose it, re-run this and set a new one.");
  console.log("  • Clear your terminal afterwards so the values aren't left on screen.\n");
}

// Only run the prompt flow when executed directly, so the derivation
// function above can be imported and tested without side effects.
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => {
    console.error("\n" + err.message);
    process.exit(1);
  });
}
